[HDR] Moving the WebView from a screen to another should change backing store formats.

mattwoodrow opened the pull request:
#### a08fe899cd50219c26ee2942bc74ba09f5322417
<pre>
[HDR] Moving the WebView from a screen to another should change backing store formats.
<a href="https://bugs.webkit.org/show_bug.cgi?id=294757">https://bugs.webkit.org/show_bug.cgi?id=294757</a>
&lt;<a href="https://rdar.apple.com/problem/153908494">rdar://problem/153908494</a>&gt;

Reviewed by Simon Fraser.

Add code to PlatformCALayerClient/GraphicsLayerClient to query the currently
supported ContentsFormats for the screen that the layer is on. Use this version,
rather than querying the values for the primary display for all layers.

Add change handling to Page (similar to deviceOrPageScaleFactorChanged) that
notifies all GraphicsLayers when this result might change.

* Source/WebCore/page/LocalFrame.cpp:
(WebCore::LocalFrame::screenContentsFormatsChanged):
* Source/WebCore/page/LocalFrame.h:
* Source/WebCore/page/Page.cpp:
(WebCore::Page::screenPropertiesDidChange):
(WebCore::Page::windowScreenDidChange):
* Source/WebCore/page/Page.h:
* Source/WebCore/platform/PlatformScreen.h:
* Source/WebCore/platform/graphics/GraphicsLayer.cpp:
(WebCore::GraphicsLayer::noteScreenContentsFormatsChangedIncludingDescendants):
* Source/WebCore/platform/graphics/GraphicsLayer.h:
(WebCore::GraphicsLayer::screenContentsFormatsChanged):
* Source/WebCore/platform/graphics/GraphicsLayerClient.h:
(WebCore::GraphicsLayerClient::screenContentsFormats const):
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp:
(WebCore::GraphicsLayerCA::createPlatformCALayer):
(WebCore::GraphicsLayerCA::ensureStructuralLayer):
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.h:
* Source/WebCore/platform/graphics/ca/PlatformCALayer.h:
* Source/WebCore/platform/graphics/ca/PlatformCALayer.mm:
(WebCore::PlatformCALayer::contentsFormatForLayer):
* Source/WebCore/platform/graphics/ca/PlatformCALayerClient.h:
* Source/WebCore/platform/graphics/ca/TileCoverageMap.cpp:
(WebCore::TileCoverageMap::screenContentsFormats const):
* Source/WebCore/platform/graphics/ca/TileCoverageMap.h:
* Source/WebCore/platform/graphics/ca/TileGrid.cpp:
(WebCore::TileGrid::screenContentsFormats const):
* Source/WebCore/platform/graphics/ca/TileGrid.h:
* Source/WebCore/platform/ios/PlatformScreenIOS.mm:
(WebCore::screenSupportsHighDynamicRange):
* Source/WebCore/platform/mac/PlatformScreenMac.mm:
(WebCore::screenSupportsHighDynamicRange):
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::screenContentsFormats const):
* Source/WebCore/rendering/RenderLayerBacking.h:
* Source/WebCore/rendering/RenderLayerCompositor.cpp:
(WebCore::RenderLayerCompositor::screenContentsFormatsChanged):
* Source/WebCore/rendering/RenderLayerCompositor.h:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/GraphicsLayerCARemote.mm:
(WebKit::GraphicsLayerCARemote::createPlatformCALayer):

Canonical link: <a href="https://commits.webkit.org/296635@main">https://commits.webkit.org/296635@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/39555cda555f28bcbc6ee3fb4cd1609cfc7f6a3c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/109035 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/28696 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/19120 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/114246 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/59353 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/110998 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/29378 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/37262 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/82863 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/111983 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/23366 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/98209 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/63305 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/22772 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/16348 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/58937 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/92735 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/16394 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/117364 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/36084 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/26672 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/91877 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/36455 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/94473 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/91682 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23363 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/36593 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/14343 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/31944 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/35981 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/41495 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/35679 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/39017 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/37364 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/page/LocalFrame.h:
```diff
@@ -223,7 +223,8 @@ class LocalFrame final : public Frame {
 
     void deviceOrPageScaleFactorChanged();
-    
+    void screenContentsFormatsChanged();
```
The name is a bit abstract. Does a screen have a "contents format"? It's more like `screenSupportedContentsFormats` or `screenAppropriateContentsFormats`

smfr commented on Source/WebCore/page/Page.cpp:
```diff
@@ -1729,6 +1729,14 @@ void Page::screenPropertiesDidChange()
 
+#if HAVE(SUPPORT_HDR_DISPLAY)
+    if (m_screenSupportsHDR != screenSupportsHighDynamicRange(m_displayID)) {
+        m_screenSupportsHDR = screenSupportsHighDynamicRange(m_displayID);
```
Let's not call `screenSupportsHighDynamicRange(m_displayID)` twice.

smfr commented on Source/WebCore/page/Page.cpp:
```diff
@@ -1748,6 +1756,14 @@ void Page::windowScreenDidChange(PlatformDisplayID displayID, std::optional<Fram
 
+#if HAVE(SUPPORT_HDR_DISPLAY)
+    if (m_screenSupportsHDR != screenSupportsHighDynamicRange(m_displayID)) {
+        m_screenSupportsHDR = screenSupportsHighDynamicRange(m_displayID);
```
Ditto.

smfr replied:
We should share some code with `screenPropertiesDidChange()`.

smfr reviewed (approved):
Can we write an API test for this?

mattwoodrow committed:
[HDR] Moving the WebView from a screen to another should change backing store formats.
https://bugs.webkit.org/show_bug.cgi?id=294757
<rdar://problem/153908494>

Reviewed by Simon Fraser.

Add code to PlatformCALayerClient/GraphicsLayerClient to query the currently
supported ContentsFormats for the screen that the layer is on. Use this version,
rather than querying the values for the primary display for all layers.

Add change handling to Page (similar to deviceOrPageScaleFactorChanged) that
notifies all GraphicsLayers when this result might change.

* Source/WebCore/page/LocalFrame.cpp:
(WebCore::LocalFrame::screenContentsFormatsChanged):
* Source/WebCore/page/LocalFrame.h:
* Source/WebCore/page/Page.cpp:
(WebCore::Page::screenPropertiesDidChange):
(WebCore::Page::windowScreenDidChange):
* Source/WebCore/page/Page.h:
* Source/WebCore/platform/PlatformScreen.h:
* Source/WebCore/platform/graphics/GraphicsLayer.cpp:
(WebCore::GraphicsLayer::noteScreenContentsFormatsChangedIncludingDescendants):
* Source/WebCore/platform/graphics/GraphicsLayer.h:
(WebCore::GraphicsLayer::screenContentsFormatsChanged):
* Source/WebCore/platform/graphics/GraphicsLayerClient.h:
(WebCore::GraphicsLayerClient::screenContentsFormats const):
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp:
(WebCore::GraphicsLayerCA::createPlatformCALayer):
(WebCore::GraphicsLayerCA::ensureStructuralLayer):
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.h:
* Source/WebCore/platform/graphics/ca/PlatformCALayer.h:
* Source/WebCore/platform/graphics/ca/PlatformCALayer.mm:
(WebCore::PlatformCALayer::contentsFormatForLayer):
* Source/WebCore/platform/graphics/ca/PlatformCALayerClient.h:
* Source/WebCore/platform/graphics/ca/TileCoverageMap.cpp:
(WebCore::TileCoverageMap::screenContentsFormats const):
* Source/WebCore/platform/graphics/ca/TileCoverageMap.h:
* Source/WebCore/platform/graphics/ca/TileGrid.cpp:
(WebCore::TileGrid::screenContentsFormats const):
* Source/WebCore/platform/graphics/ca/TileGrid.h:
* Source/WebCore/platform/ios/PlatformScreenIOS.mm:
(WebCore::screenSupportsHighDynamicRange):
* Source/WebCore/platform/mac/PlatformScreenMac.mm:
(WebCore::screenSupportsHighDynamicRange):
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::screenContentsFormats const):
* Source/WebCore/rendering/RenderLayerBacking.h:
* Source/WebCore/rendering/RenderLayerCompositor.cpp:
(WebCore::RenderLayerCompositor::screenContentsFormatsChanged):
* Source/WebCore/rendering/RenderLayerCompositor.h:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/GraphicsLayerCARemote.mm:
(WebKit::GraphicsLayerCARemote::createPlatformCALayer):

Canonical link: https://commits.webkit.org/296635@main
